### ISSUE TITLE
sundials build issue with +rocm

### ISSUE THREAD

**Issue by balay** [xsdk-project/xsdk-issues#239]
```
make[2]: Leaving directory '/scratch/balay/spack/spack-stage/spack-stage-sundials-6.6.1-63n33g3wayum7j2l62rkp6xw2simijok/spack-build-63n33g3'
In file included from /scratch/balay/spack/spack-stage/spack-stage-sundials-6.6.1-63n33g3wayum7j2l62rkp6xw2simijok/spack-src/src/sunmatrix/magmadense/sunmatrix_magmadense.cpp:19:
In file included from /scratch/balay/spack/spack-stage/spack-stage-sundials-6.6.1-63n33g3wayum7j2l62rkp6xw2simijok/spack-src/include/sunmatrix/sunmatrix_magmadense.h:33:
In file included from /scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/magma-2.7.1-jiafhw6lufkrl3zya76zczclo62vv5zp/include/magma_v2.h:22:
In file included from /scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/magma-2.7.1-jiafhw6lufkrl3zya76zczclo62vv5zp/include/magmablas.h:12:
In file included from /scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/magma-2.7.1-jiafhw6lufkrl3zya76zczclo62vv5zp/include/magma_copy.h:12:
/scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/magma-2.7.1-jiafhw6lufkrl3zya76zczclo62vv5zp/include/magma_types.h:150:14: fatal error: 'hipblas/hipblas.h' file not found
    #include <hipblas/hipblas.h>
             ^~~~~~~~~~~~~~~~~~~
```

[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/12908467/spack-build-out.txt)

**Comment by balay** [comment]
For now trying:
```patch
diff --git a/var/spack/repos/builtin/packages/sundials/package.py b/var/spack/repos/builtin/packages/sundials/package.py
index f424a523f1..12cf249d6b 100644
--- a/var/spack/repos/builtin/packages/sundials/package.py
+++ b/var/spack/repos/builtin/packages/sundials/package.py
@@ -231,6 +231,7 @@ class Sundials(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("raja", when="+raja")
     depends_on("raja+cuda", when="+raja +cuda")
     depends_on("raja+rocm", when="+raja +rocm")
+    depends_on("hipblas", when="+magma +rocm")
 
     # External libraries
     depends_on("caliper", when="+caliper")
@@ -454,11 +455,13 @@ def cmake_args(self):
 
         # Building with MAGMA
         if "+magma" in spec:
-            args.extend([define("ENABLE_MAGMA", True), define("MAGMA_DIR", spec["magma"].prefix)])
             if "+cuda" in spec:
+                args.extend([define("ENABLE_MAGMA", True), define("MAGMA_DIR", spec["magma"].prefix)])
                 args.extend([define("SUNDIALS_MAGMA_BACKENDS", "CUDA")])
             if "+rocm" in spec:
                 args.extend([define("SUNDIALS_MAGMA_BACKENDS", "HIP")])
+                args.extend([define("MAGMA_INCLUDE_DIR",';'.join([spec["magma"].prefix.include,spec["hipblas"].prefix.include]))])
+                args.extend([define("MAGMA_LIBRARY_DIR", spec["magma"].prefix.lib)])
 
         # Building with PETSc
         if "+petsc" in spec:
```

BTW: Does magma need to include hipblas.h in its public include file?

**Comment by balay** [comment]
ops that's buggy ['ENABLE_MAGMA' got removed]. For now - I'm using:
```patch
diff --git a/var/spack/repos/builtin/packages/sundials/package.py b/var/spack/repos/builtin/packages/sundials/package.py
index f424a523f1..850f110b3a 100644
--- a/var/spack/repos/builtin/packages/sundials/package.py
+++ b/var/spack/repos/builtin/packages/sundials/package.py
@@ -257,6 +257,8 @@ class Sundials(CMakePackage, CudaPackage, ROCmPackage):
     depends_on("hypre@:2.22.0+mpi~int64", when="@:5.7.0 +hypre ~int64")
     depends_on("hypre@:2.22.0+mpi+int64", when="@:5.7.0 +hypre +int64")
     depends_on("magma", when="+magma")
+    depends_on("hipblas", when="+magma +rocm")
+    depends_on("hipsparse", when="+magma +rocm")
     depends_on("petsc+mpi", when="+petsc")
     depends_on("suite-sparse", when="+klu")
     depends_on("superlu-dist@6.1.1:", when="@:5.4.0 +superlu-dist")
@@ -454,11 +456,14 @@ def cmake_args(self):
 
         # Building with MAGMA
         if "+magma" in spec:
-            args.extend([define("ENABLE_MAGMA", True), define("MAGMA_DIR", spec["magma"].prefix)])
+            args.extend([define("ENABLE_MAGMA", True)])
             if "+cuda" in spec:
+                args.extend([define("MAGMA_DIR", spec["magma"].prefix)])
                 args.extend([define("SUNDIALS_MAGMA_BACKENDS", "CUDA")])
             if "+rocm" in spec:
                 args.extend([define("SUNDIALS_MAGMA_BACKENDS", "HIP")])
+                args.extend([define("MAGMA_INCLUDE_DIR", [spec["magma"].prefix.include, spec["hipblas"].prefix.include, spec["hipsparse"].prefix.include])])
+                args.extend([define("MAGMA_LIBRARY", spec["magma"].libs)])
 
         # Building with PETSc
         if "+petsc" in spec:
```

**Comment by luszczek** [comment]
Initially, looks like Spack problem because MAGMA's `magma/package.py` says:

```python
depends_on("hipblas", when="+rocm")
depends_on("hipsparse", when="+rocm")
```

So the question is why SUNDIALS has to say the exact same thing to get HIP libraries included. Maybe @stomov have seen a similar issue before with other downstream libraries.

**Comment by balay** [comment]
> So the question is why SUNDIALS has to say the exact same thing to get HIP

Its the  includes - because the hip inclues are exposed via magma public include files. Do they need to be exposed in public includes - or is it needed only for magma sources? If so  - perhaps that should be in a private include file - used only by magma sources?

**Comment by balay** [comment]
I'll also note: my rocm install is with spack - so it installs hip,hipblas,hipsparse etc in separate locations [hence triggers this error].

```console
balay@petsc-gpu-02:/scratch/balay/spack$ ./bin/spack find -p rocblas rocsparse
-- linux-ubuntu22.04-zen4 / gcc@11.4.0 --------------------------
rocblas@5.5.1    /scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/rocblas-5.5.1-o5dfli2vi3kyracxu2ogzhxe5qx4mt6q
rocsparse@5.5.1  /scratch/balay/spack/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/rocsparse-5.5.1-ent62v3xvgfmbxh3ktqrnbzetxx2rfug
==> 2 installed packages
balay@petsc-gpu-02:/scratch/balay/spack$ 
```
Alternative is to use a preinstalled rocm - like /opt/rocm-5.4.3 - and list all rocm pkgs in ~/.spack/package.yml. Here this issue might not come up as the same -I//opt/rocm-5.4.3/include works for all these rocm pkgs..

**Comment by stomov** [comment]
Regarding SUNDAILS using MAGMA with rocm, I assume if on AMD GPUs they also use directly some of the AMD math libraries, so probably that's why listing the dependencies? One challenge for making clean include files in MAGMA are the argument types - in particular complex types and half that every vendor defines differently (AMD had different complex types even for different libraries) - and sometimes we use those. 
Regarding the different include directories, this reminds me the CUDA installation on Perlmutter - libraries were at different locations and cmake could not find them. We did hacks in the packages but at some point somebody introduced change in cmake to do
find_package( CUDAToolkit )
and with that
include_directories( ${CUDAToolkit_INCLUDE_DIRS} )
was just working. Not sure if something like this is available for HIP. If yes, we can change magma.

**Comment by balos1** [comment]
> Alternative is to use a preinstalled rocm - like /opt/rocm-5.4.3 - and list all rocm pkgs in ~/.spack/package.yml. Here this issue might not come up as the same -I//opt/rocm-5.4.3/include works for all these rocm pkgs..

I was wondering why we had never seen this before, but we always test with the preinstalled rocm so it makes sense. MAGMA.  I think the spack change is reasonable for now, but I think the ideal way for this to get fixed would be for the MAGMA exported CMake targets to include hipblas, hipsparse, etc. However, I am not sure what exactly the current state of hip targets and CMake supports.

**Comment by balay** [comment]
Note: same issue with using `magma` from `heffte` [and similar fix]

https://github.com/xsdk-project/xsdk-issues/issues/228

cc: @mkstoyanov

**Comment by mkstoyanov** [comment]
@balos1 I think it's s spack issue and not CMake. MAGMA doesn't export any dependencies, in heFFTe's CMake we create the MAGMA import and search for the dependencies ourselves. However, the search fails when using spack, which I think is yet another failure for spack to handle the dependencies (transitively in this case). In fact, I've had very bad experience with spack over the years, and this month again I have multiple installations of rocblas and rocfft, which take hours to build.

Using an external ROCm will always load all ROCm libraries into the spack environment, so the search will work fine.

**Comment by balay** [comment]
Its more than a "spack" issue
- a - spack needs to provide mechanism to discover such optional dependencies of dependent pkgs
- b - also the package needs a mechanism to import these dependencies - and use them

Here - I had to hack `MAGMA_INCLUDE_DIR` for (b). Note not all pkgs are with cmake - so any cmake fix won't help non-cmake pkgs..

The way I deal with this in petsc [for (b) - where configure does handle (a)] is construct `CPPFLAGS=-Ip1 -Ip2` [instead of PKG1_INCLUDE=p1, PKG2_INCLIDE=p2 etc..]. And similarly for `LDFLAGS=--Lp1 -lp1 -Lp2 -lp2'` etc instead of PKG1_LIBS= PKG2_LIBS etc] - but this requires pkgs to support CPPFLAGS/LDFLAGS configure options..

**Comment by balay** [comment]
Also note: similar issue exists with building/using static libraries - hence we couldn't work on `xsdk+static` variant.

Shared libraries avoid this problem a bit [for the libraries]. And its good to avoid exposing the dependencies via "includes"/public-API - if possible.

> One challenge for making clean include files in MAGMA are the argument types - in particular complex types and half that every vendor defines differently

Yeah if the hip/rocm "datatypes" are used in MAGMA API [i.e function prototypes] - its not possible to avoid exposing it in MAGMA public includes.

**Comment by mkstoyanov** [comment]
Exporting-importing has to be done "correctly", but the "correct" way is different for different cases. I don't know how this is done with autotools, I have seen package files that contain meta-data and even CMake can work with those, but I'm far from qualified to comment.

MAGMA uses CMake and does not have a correct export mechanism for the transitive dependencies, i.e., when a package uses MAGMA though CMake, the package has to manually "know" which libraries are used by MAGMA and find those through the CMake mechanisms. This is a MAGMA issue. However, the "correct" CMake solution of adding exports to the MAGMA CMake will still fail in spack.

The issue with spack is that it needs to load the transitive dependencies in the CMake search paths. The transitive dependencies need to be loaded. Using shared libraries can help hide some those issues at compile time, but the headers are still needed. Shared and static libraries will fail at runtime (or testing), if the correct dependencies are not loaded.

**Comment by pghysels** [comment]
Related issue on Perlmutter with `sundials +cuda`.
It cannot find `cublas` or `cusparse` (needed for `magma`), which on Perlmutter are installed in a separate folder from the core cuda libraries.
[spack-build-out.txt](https://github.com/xsdk-project/xsdk-issues/files/12921772/spack-build-out.txt)

**Comment by mkstoyanov** [comment]
The work-around that I am using is to add the transitive dependencies manually, e.g.,
```
   depends_on("rocblas@5.0:", when="+magma+rocm")
    depends_on("rocsparse@5.0:", when="+magma+rocm")
    depends_on("hipblas@5.0:", when="+magma+rocm")
    depends_on("hipsparse@5.0:", when="+magma+rocm")
```
You can try the same for cuda.

**Comment by luszczek** [comment]
I think @stomov had special instructions for Perlmutter's CUDA configuration when building MAGMA. These may help on HIP/ROCm systems as well. Although CUDA usually still have the stubs and the real libraries that get used at different stages: compilation and runtime.

**Comment by balos1** [comment]
This has been fixed with sundials 6.6.2
